sys/net/routing/ng_rpl/ng_rpl_dodag.c

```diff
@@ -0,0 +1,384 @@
+    for (uint8_t i = 0; i < NG_RPL_INSTANCES_NUMOF; ++i) {
+        /* save position to the first unused instance */
+        if ((ng_rpl_instances[i].state == 0) && first) {
+            *inst = &ng_rpl_instances[i];
+            first = false;
+            continue;
+        }
+        else if ((ng_rpl_instances[i].state != 0) && (ng_rpl_instances[i].id == instance_id)) {
+            DEBUG("Instance with id %d exists\n", instance_id);
+            *inst = &ng_rpl_instances[i];
+            return false;
+        }
+    }
+
+    if (*inst != NULL) {
+        (*inst)->id = instance_id;
+        (*inst)->state = 1;
+        (*inst)->maxrankincrease = NG_RPL_DEFAULT_MAX_RANK_INCREASE;
+        (*inst)->minhoprankincrease = NG_RPL_DEFAULT_MIN_HOP_RANK_INCREASE;
+        (*inst)->dodags = NULL;
+        return true;
+    }
+
+    /* no space available to allocate a new instance */
+    DEBUG("Could not allocate a new RPL instance\n");
+    *inst = NULL;
+    return false;
+}
+
+bool ng_rpl_instance_remove_by_id(uint8_t instance_id)
+{
+    for(uint8_t i = 0; i < NG_RPL_INSTANCES_NUMOF; ++i) {
+        if (ng_rpl_instances[i].id == instance_id) {
+            ng_rpl_dodag_t *elt, *tmp;
+            LL_FOREACH_SAFE(ng_rpl_instances[i].dodags, elt, tmp) {
+                ng_rpl_dodag_remove(elt);
+            }
+            memset(&ng_rpl_instances[i], 0, sizeof(ng_rpl_instance_t));
+            return true;
+        }
+    }
+    return false;
+}
+
+bool ng_rpl_instance_remove(ng_rpl_instance_t *inst)
+{
+    ng_rpl_dodag_t *elt, *tmp;
+    LL_FOREACH_SAFE(inst->dodags, elt, tmp) {
+        ng_rpl_dodag_remove(elt);
+    }
+    memset(inst, 0, sizeof(ng_rpl_instance_t));
+    return true;
+}
+
+ng_rpl_instance_t *ng_rpl_instance_get(uint8_t instance_id)
+{
+    for (uint8_t i = 0; i < NG_RPL_INSTANCES_NUMOF; ++i) {
+        if (ng_rpl_instances[i].id == instance_id) {
+            return &ng_rpl_instances[i];
+        }
+    }
+    return NULL;
+}
+
+bool ng_rpl_dodag_add(ng_rpl_instance_t *instance, ng_ipv6_addr_t *dodag_id, ng_rpl_dodag_t **dodag)
+{
+    if ((instance == NULL) || instance->state == 0) {
+        DEBUG("Instance is NULL or unused\n");
+        return NULL;
+    }
+
+    *dodag = NULL;
+    bool first = true;
+    for (uint8_t i = 0; i < NG_RPL_DODAGS_NUMOF; ++i) {
+        /* save position to the first unused instance */
+        if ((ng_rpl_dodags[i].state == 0) && first) {
+            *dodag = &ng_rpl_dodags[i];
+            first = false;
+            continue;
+        }
+        else if ((ng_rpl_dodags[i].state != 0) &&
+                (ng_rpl_dodags[i].instance->id == instance->id) &&
+                ng_ipv6_addr_equal(&ng_rpl_dodags[i].dodag_id, dodag_id)) {
+            DEBUG("DODAG with id: %s does exist\n", ng_ipv6_addr_to_str(addr_str, dodag_id,
+                        sizeof(addr_str)));
+            *dodag = &ng_rpl_dodags[i];
+            return false;
+        }
+    }
+
+    if (*dodag != NULL) {
```

Comment:
`dodag` is of type `ng_rpl_dodag_t **`. I need to dereference it in order to check if it contains a valid dodag address or not